lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+                             originalStride));
+  auto subview = rewriter.create<memref::SubViewOp>(
+      op.getLoc(), dyn_cast<MemRefType>(subviewType), expandShape.getResult(),
+      offsets, sizes, strides);
+  originalSubviewOp->replaceAllUsesWith(subview);
+}
+
+static LogicalResult insertParallelLoop(ForallOp op, unsigned mTileShape,
+                                        unsigned nTileShape) {
+  xsmm::BrgemmOp brgemmOp = NULL;
+  OpBuilder b(op);
+  IRRewriter rewriter(b.getContext());
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++)
+    if (dyn_cast<xsmm::BrgemmOp>(oper)) {
+      brgemmOp = dyn_cast<xsmm::BrgemmOp>(oper);
+      break;
+    }
+  if (brgemmOp == NULL)
+    return rewriter.notifyMatchFailure(op, "require brgemm op");
+
+  if (mTileShape == 0 || nTileShape == 0)
+    return rewriter.notifyMatchFailure(op, "require tile shape to not be zero");
+
+  ArrayRef<int64_t> mShape;
+  auto mDefiningOp = brgemmOp.getOperand(1).getDefiningOp();
+  if (isa<memref::SubViewOp>(mDefiningOp))
+    mShape =
+        dyn_cast<ShapedType>(mDefiningOp->getOperand(0).getType()).getShape();
+  else
+    mShape =
+        dyn_cast<ShapedType>(mDefiningOp->getResult(0).getType()).getShape();
+
+  ArrayRef<int64_t> nShape;
+  auto nDefiningOp = brgemmOp.getOperand(2).getDefiningOp();
+  if (isa<memref::SubViewOp>(nDefiningOp))
+    nShape =
+        dyn_cast<ShapedType>(nDefiningOp->getOperand(0).getType()).getShape();
+  else
+    nShape =
+        dyn_cast<ShapedType>(nDefiningOp->getResult(0).getType()).getShape();
+
+  ArrayRef<int64_t> kShape;
+  auto kDefiningOp = brgemmOp.getOperand(3).getDefiningOp();
+  if (isa<memref::SubViewOp>(kDefiningOp))
+    kShape =
+        dyn_cast<ShapedType>(kDefiningOp->getOperand(0).getType()).getShape();
+  else
+    kShape =
+        dyn_cast<ShapedType>(kDefiningOp->getResult(0).getType()).getShape();
+
+  if (mShape.size() != kShape.size())
+    return rewriter.notifyMatchFailure(
+        op, "require m tensor shape and k tensor shape to match");
+
+  SmallVector<unsigned> tileShapeM;
+  tileShapeM.push_back(mTileShape);
+  tileShapeM.push_back(mShape[0] / mTileShape);
+
+  SmallVector<unsigned> tileShapeN;
+  tileShapeN.push_back(nTileShape);
+  tileShapeN.push_back(nShape[0] / nTileShape);
+
+  // Validate the input tile sizes against the operand shapes
+  long multipleM = 1;
+  for (size_t i = 0; i < tileShapeM.size(); i++)
+    multipleM = multipleM * tileShapeM[i];
+
+  if (mShape[0] != multipleM)
+    return rewriter.notifyMatchFailure(
+        op, "require m tile shape to match tensor shape");
+
+  long multipleN = 1;
+  for (size_t i = 0; i < tileShapeN.size(); i++)
+    multipleN = multipleN * tileShapeN[i];
+
+  if (nShape[0] != multipleN)
+    return rewriter.notifyMatchFailure(
+        op, "require n tile shape to match tensor shape");
+
+  if ((multipleM * multipleN) != (kShape[0] * kShape[1]))
```

Comment:
This is assuming the `kShape` is 2D. No more, no less. If that's the requirement, then we need to verify that this is the case for m, n and k.